### PR TITLE
Remediate [BVR-06M] Insufficient Protection of Contract Assets

### LINKED ISSUE #743: [BVR-06M] Insufficient Protection of Contract Assets
## BVR-06M: Insufficient Protection of Contract Assets

| Type | Severity | Location |
| ---- | -------- | -------- |
| [Logical Fault](/appendix/finding-types#logical-fault) |  | [BosonVoucher.sol:L536-L543](https://github.com/bosonprotocol/boson-protocol-contracts/blob/02a4d2ff04832a4607674ba37708d4f52cdac4fa/contracts/protocol/clients/voucher/BosonVoucher.sol#L536-L543) |

### Description:

The [`BosonVoucherBase::callExternalContract`](https://github.com/bosonprotocol/boson-protocol-contracts/blob/02a4d2ff04832a4607674ba37708d4f52cdac4fa/contracts/protocol/clients/voucher/BosonVoucher.sol#L531-L546) is insecure as it permits the contract to perform arbitrary calls, potentially compromising the assets it is in possession of. A short list of disallowed `selector` values has been specified, however, it is insufficient if the contract is expected to not reduce its balance in any way.

### Impact:

It is presently possible to compromise funds held within the `BosonVoucher` despite the security measures in place within [`BosonVoucherBase::callExternalContract`](https://github.com/bosonprotocol/boson-protocol-contracts/blob/02a4d2ff04832a4607674ba37708d4f52cdac4fa/contracts/protocol/clients/voucher/BosonVoucher.sol#L531-L546).

### Example:

```solidity
function callExternalContract(address _to, bytes calldata _data) external payable onlyOwner returns (bytes memory) {
    require(_to != address(0), INVALID_ADDRESS);

    // Prevent invocation of functions that would allow transfer of tokens from this contract
    bytes4 selector = bytes4(_data[:4]);
    require(
        selector != IERC20.transfer.selector &&
            selector != IERC20.approve.selector &&
            selector != IERC20.transferFrom.selector &&
            selector != DAI.push.selector &&
            selector != DAI.move.selector,
        FUNCTION_NOT_ALLOWLISTED
    );

    return _to.functionCallWithValue(_data, msg.value, FUNCTION_CALL_NOT_SUCCESSFUL);
}
```

### Recommendation:

We advise either the `selector` list to be made a `mapping` that can be maintained by the Boson Protocol team as the current solution is insufficient in preventing token transfers.

As tangible examples, an `ERC20::burn` call can be made on the `DAI` token, an `ERC20Like::increaseAllowance` call can be made on the `USDC` token, and `ERC20Like::increaseApproval` as well as `ERC20Like::increaseAllowance` functions exist on multiple tokens.

Depending on the intended purpose of the [`BosonVoucherBase::callExternalContract`](https://github.com/bosonprotocol/boson-protocol-contracts/blob/02a4d2ff04832a4607674ba37708d4f52cdac4fa/contracts/protocol/clients/voucher/BosonVoucher.sol#L531-L546) function, multiple robust solutions can be employed. We will detail a few approaches in brief that the Boson Protocol can follow below:

- Implement a "target" whitelist in case `BosonVoucher` instances are expected to interact with a small subset of contracts (i.e. [Seaport based on the PR of the changes](https://github.com/bosonprotocol/boson-protocol-contracts/pull/571)).
- Implement a blanket asset protection measure by evaluating whether the `_to` address supports the `ERC20::balanceOf` function. As almost all [EIP-20](https://eips.ethereum.org/EIPS/eip-20) tokens require direct interaction to transfer tokens or approve, a successful `ERC20::balanceOf` invocation would indicate the target `_to` address is a token and should not be interacted with.
- Maintain a protocol-wide `selector` blacklist that uses a `mapping` entry that can be updated in the future as more signatures need to be blacklisted.

To note, the contract seems to solely protect [EIP-20](https://eips.ethereum.org/EIPS/eip-20) assets. If [EIP-721](https://eips.ethereum.org/EIPS/eip-721) or [EIP-1155](https://eips.ethereum.org/EIPS/eip-1155) assets need to be protected as well (which we believe to be the case), all the aforementioned solutions would need to be adapted accordingly to support those two standards on top (i.e. `ERC721::ownerOf` instead of `ERC20::balanceOf`).

### PR DESCRIPTION
**Pull request by zajck**
Fix #743

### CODE REVIEW

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 1]
```diff
@@ -527,7 +527,8 @@ contract BosonVoucherBase is IBosonVoucher, BeaconClientBase, OwnableUpgradeable
      * - _to is zero address
      * - call to external contract fails
      * - caller is not the owner
-     * - caller tries to call ERC20 method that would allow transfer of tokens from this contract
+     * - _to is a contract that represents some assets (all contracts that imeplement `balanceOf` method, including ERC20 and ERC721)
```
Comment (albertfolch-redeemeum) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:530]:
```suggestion
     * - _to is a contract that represents some assets (all contracts that implement `balanceOf` method, including ERC20 and ERC721)
```

File: contracts/interfaces/clients/IBosonVoucher.sol [hunk 2]
```diff
@@ -216,7 +216,7 @@ interface IBosonVoucher is IERC721Upgradeable, IERC721MetadataUpgradeable, IERC7
      * - _to is zero address
      * - call to external contract fails
      * - caller is not the owner
-     * - caller tries to call ERC20 method that would allow transfer of tokens from this contract
+     * - _to is a contract that represents some assets (all contracts that imeplement `balanceOf` method, including ERC20 and ERC721)
```
Comment (albertfolch-redeemeum) [inline contracts/interfaces/clients/IBosonVoucher.sol:219]:
```suggestion
     * - _to is a contract that represents some assets (all contracts that implement `balanceOf` method, including ERC20 and ERC721)
```

File: test/protocol/clients/BosonVoucherTest.js [hunk 3]
```diff
@@ -2141,49 +2141,85 @@ describe("IBosonVoucher", function () {
         ).to.be.reverted;
       });
 
-      it("Owner tries to invoke method to transfer funds", async function () {
-        const erc20 = await getContractFactory("Foreign20");
+      it("Owner tries to interact with contract with assets", async function () {
+        // const erc20 = await getContractFactory("Foreign20");
```
Comment (albertfolch-redeemeum) [inline test/protocol/clients/BosonVoucherTest.js:2145]: remove 
```suggestion
```

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 4]
```diff
@@ -538,16 +539,12 @@ contract BosonVoucherBase is IBosonVoucher, BeaconClientBase, OwnableUpgradeable
...
-            selector != IERC20.transfer.selector &&
-                selector != IERC20.approve.selector &&
-                selector != IERC20.transferFrom.selector &&
-                selector != DAI.push.selector &&
-                selector != DAI.move.selector,
-            FUNCTION_NOT_ALLOWLISTED
-        );
+        // Check if _to supports `balanceOf` method and revert if it does
+        // This works with all contracts that implement this method even if they don't necessary implement ERC20 interface
+        uint256 balance;
```
Comment (levalleux-ludo) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:544]: Useless variable `balance`
```suggestion
```

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 5]
```diff
@@ -538,16 +539,12 @@ contract BosonVoucherBase is IBosonVoucher, BeaconClientBase, OwnableUpgradeable
...
+        try IERC20(_to).balanceOf(address(this)) returns (uint256 _balance) {
```
Comment (levalleux-ludo) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:545]: Useless variable `_balance`
```suggestion
        try IERC20(_to).balanceOf(address(this)) returns (uint256) {
```

**Review by levalleux-ludo — CHANGES_REQUESTED** [review]
2 small suggestions

**Review by levalleux-ludo — APPROVED** [review]
LGTM

**Review by albertfolch-redeemeum — APPROVED** [review]

**Pull request merged**